Indirect dependency to OpenSSL library cannot be represented by CMake for some situation

The openssl_posix target has the following dependencies:

```
target_link_libraries( openssl_posix
                       PUBLIC
                          sockets_posix
                       PRIVATE
                          # This variable is set by the built-in FindOpenSSL.cmake
                          # and contains the path to the actual library.
                          ${OPENSSL_LIBRARIES}
                          # SSL uses Threads and on some platforms require
                          # explicit linking.
                          Threads::Threads
                          # SSL uses Dynamic Loading and on some platforms
                          # requires explicit linking.
                          ${CMAKE_DL_LIBS} )
```

as we can see, the OpenSSL libraries (libcrypto.so, libssl.so) are put in the PRIVATE type dependency.

For normal situation there won't be any issue when doing native build or cross-build.
For some situation, if the OpenSSL libraries are not in their standard location (like /lib, /usr/lib, etc) the following error is seen when linking the http_demo_basic_tls demo executable:

```
[ 66%] Linking C executable ../../../bin/http_demo_basic_tls
/tmp/aarch64-linux/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/../../../../aarch64-buildroot-linux-gnu/bin/ld: warning: libssl.so.1.1, needed by ../../../lib/libopenssl_posix.so, not found (try using -rpath or -rpath-link)
/tmp/aarch64-linux/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/../../../../aarch64-buildroot-linux-gnu/bin/ld: warning: libcrypto.so.1.1, needed by ../../../lib/libopenssl_posix.so, not found (try using -rpath or -rpath-link)
/tmp/aarch64-linux/bin/../lib/gcc/aarch64-buildroot-linux-gnu/9.3.0/../../../../aarch64-buildroot-linux-gnu/bin/ld: ../../../lib/libopenssl_posix.so: undefined reference to `SSL_shutdown@OPENSSL_1_1_0'
...
```

http_demo_basic_tls only depends on library clock_posix and openssl_posix.

```
target_link_libraries(
    ${DEMO_NAME}
    PRIVATE
        clock_posix
        openssl_posix
)
```

So only clock_posix and openssl_posix library will appear in the linker's command line when linking http_demo_basic_tls, but since openssl_posix has indirect dependency to OpenSSL libraries itself, the linker will try to do symbol resolution at link-time. If the OpenSSL libraries are not in their normal location, the linker fails to locate them and reports link error, unless "-rpath-link" command-line option is given to the linker.

We can reproduce such build failure in Linux, by moving OpenSSL libraries (libcrypto.so, libssl.so) from their usual location (/usr/lib) to somewhere else, for example, placing it to /home/openssl, and pass -DCMAKE_LIBRARY_PATH=/home/openssl -DOPENSSL_ROOT_DIR=/home/openssl to CMake.

This issue is likely related to the following CMake issue:
https://gitlab.kitware.com/cmake/cmake/-/issues/22953

A proved working workaround is to change OpenSSL library dependency type from PRIVATE to PUBLIC.

--- a/platform/posix/transport/CMakeLists.txt
+++ b/platform/posix/transport/CMakeLists.txt
@@ -29,10 +29,10 @@ add_library( openssl_posix
 target_link_libraries( openssl_posix
                        PUBLIC
                           sockets_posix
-                       PRIVATE
                           # This variable is set by the built-in FindOpenSSL.cmake
                           # and contains the path to the actual library.
                           ${OPENSSL_LIBRARIES}
+                       PRIVATE
                           # SSL uses Threads and on some platforms require
                           # explicit linking.
                           Threads::Threads

